### DISCUSSION TITLE
How to create a fallback command if the command is not found?

### DISCUSSION THREAD

**Discussion by manoraj** [oclif/oclif#1494] [Q&A] [answered]
**Is your feature request related to a problem? Please describe.**
I am creating a CLI (my-cli), and it has a `run` command and some others. As `run` is the most used command, I want to make it the run this command when the command is not found.

For example, I want to run `my-cli run test` with `my-cli test` (provided test is not a command)

Real world examples: `yarn run test` and `yarn test` have same behaviour

**Describe the solution you'd like**
I want to configure the fallback command if the command is not found

**Describe alternatives you've considered**
I have tried alias/default and it didn't work

**Additional context**

**Comment by manoraj** [comment]
I kind of implemented it, but I want to check if there is a better approach.

```
export async function run({ dev }: { dev: boolean } = { dev: false }) {
	const config = await Config.load({
		...(dev ? getDevConfig() : {}),
		root: resolve(__dirname, ".."),
		enablePerf: true,
	});
	const commands = config.commandIDs;
	const args = process.argv.slice(2);
	const command = args[0];
	// @ts-expect-error - fallbackCommand is not in the Config type
	const fallbackCommand: string = config.pjson.oclif.fallbackCommand || "";

	if (fallbackCommand && command) {
		// Check if the first argument is a command or not. If not, then insert the fallback command id into the argv array.
		if (command !== "--help" && command !== "-h" && !commands.includes(command)) {
			args.unshift(fallbackCommand);
		}
	}
	return execute({
		args,
		development: dev,
		loadOptions: config,
	});
}
```

**Reply by mdonnalley** [reply, marked as answer]
That looks like it would work. You could also consider implementing the `command_not_found` hook. That hook will run anytime a provided command isn't found. Inside the hook you can use the `runCommand` method on `config` to run the desired command.

Something like this:

```typescript
import {Hook} from '@oclif/core'

const hook: Hook.CommandNotFound = async function (opts) {
  await opts.config.runCommand('run', opts.argv)
}
```

https://oclif.io/docs/hooks/#lifecycle-events
https://github.com/oclif/core/blob/4.0.14/src/interfaces/hooks.ts#L25
example implementation of `command_not_found`: https://github.com/oclif/plugin-not-found/

**Comment by manoraj** [comment]
Thanks for the suggestion. One observation though

```
import {Hook} from '@oclif/core'

const hook: Hook.CommandNotFound = async function (opts) {
  await opts.config.runCommand('run', opts.argv)
}
```

In the above code, `opts.argv` is not the actual `process.argv`. Is there a way to get the actual argv from the opts instead of relying `process.argv`? Otherwise, `await opts.config.runCommand('run', [...process.argv])` works fine.

**Reply by mdonnalley** [reply]
`opts.argv` is the equivalent of `process.argv.slice(2)`, which will remove the node executable and the js file being executed.

So for example, the process.argv might look like this:
```
 [
  '/Users/mdonnalley/.volta/tools/image/node/20.13.1/bin/node',
  '/Users/mdonnalley/.volta/tools/image/packages/@salesforce/cli/bin/sf',
  'version',
  '--versbose'
]
```

`opts.argv` will give you `['version', '--verbose']`

But if you need the entire `process.argv`, there's no harm in using that the way you described

**Comment by sergei-dyshel** [comment]
Worth noting that `command_not_found` hook does NOT fire if there are no arguments/flags are given at all. In this case the help will be printed. 

I couldn't find a "canonical" way to solve this, so used this hack to modify `process.argv` before oclif starts parsing, e.g.:
```
if (process.argv.length === 2) process.argv.push('<default-command-name>')
```
